### PR TITLE
drivers: uart_stm32: add DCache support in async DMA mode by using nocache memory regions

### PR DESCRIPTION
**Pull request by ajarmouni-st**
Adapt the stm32 uart driver to verify if DMA buffers, used in async mode, are located in noncacheable memory when DCache is activated, in order to avoid cache coherency issues.
Modify `tests/drivers/uart_async_api` test to run with DCache enabled by putting DMA buffers in a nocache memory region.

### CODE REVIEW

File: tests/drivers/uart/uart_async_api/Kconfig [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+#
+# SPDX-License-Identifier: Apache-2.0
+
+mainmenu "UART Async Test"
+
+source "Kconfig.zephyr"
+
+if DCACHE
+
+config DT_DEFINED_NOCACHE
```
Comment (GeorgeCGV) [inline tests/drivers/uart/uart_async_api/Kconfig:11]: Perhaps it's time to generalize the DT NOCACHE? Considering its usage here, in the SPI tests, and potentially by applications.

For example, by introducing related configs in https://github.com/zephyrproject-rtos/zephyr/blob/414b3b7efd3e927a7e68813217285b30a160d4b2/arch/Kconfig#L385

**Comment by ajarmouni-st** [comment]
CI failures are unrelated to this PR
```
INFO    - The following issues were found (showing the top 10 items):
INFO    - 1) samples/boards/nrf/nrf53_sync_rtc/sample.boards.nrf.nrf53_sync_rtc.simu on nrf5340bsim/nrf5340/cpuapp error (Cmake build failure)
INFO    - 2) samples/drivers/mbox/sample.drivers.mbox.simu on nrf5340bsim/nrf5340/cpuapp error (Cmake build failure)
INFO    - 3) samples/subsys/logging/multidomain/sample.logging.multidomain.ipc_static_vrings.simu on nrf5340bsim/nrf5340/cpuapp error (Cmake build failure)
```

File: drivers/serial/uart_stm32.c [hunk 2]
```diff
@@ -1512,6 +1545,12 @@ static int uart_stm32_async_tx(const struct device *dev,
 		return -EBUSY;
 	}
 
+#ifdef CONFIG_DCACHE
+	if (!buf_in_nocache((uintptr_t)tx_data, buf_size)) {
+		LOG_ERR("Tx buffer should be placed in a nocache memory region");
```
Comment (FRASTM) [inline drivers/serial/uart_stm32.c:1550]: should execution continue ?

Comment (ajarmouni-st) [reply drivers/serial/uart_stm32.c:1550]: Having a DMA buffer in a cached memory region does not always mean the execution won't be successful, it only tells us that there may be cache coherence issues absent proper cache coherence management, that's why I didn't want to abort the execution. But I'm not sure if this is the proper why to treat it.

Comment (ajarmouni-st) [reply drivers/serial/uart_stm32.c:1550]: Having a DMA buffer in a cached memory region does not always mean the execution won't be successful, it only tells us that there may be cache coherence issues absent proper cache coherence management, that's why I didn't want to abort the execution. But I'm not sure if this is the proper why to handle it.

File: tests/drivers/uart/uart_async_api/testcase.yaml [hunk 3]
```diff
@@ -75,3 +75,36 @@ tests:
...
+  drivers.uart.async_api.nocache_mem_dt.nucleo_f746zg:
+    harness: ztest
+    harness_config:
+      fixture: gpio_loopback
+    depends_on: gpio
+    platform_allow:
+      - nucleo_f746zg
+    extra_args:
+      - DTC_OVERLAY_FILE="boards/nucleo_f746zg.overlay;boards/nucleo_f746zg_nocachemem.overlay"
+      - EXTRA_CONF_FILE=stm32_nocache-mem-dt.conf
```
Comment (erwango) [inline tests/drivers/uart/uart_async_api/testcase.yaml:100]: s/'-'/'_'

File: tests/drivers/uart/uart_async_api/testcase.yaml [hunk 4]
```diff
@@ -75,3 +75,36 @@ tests:
...
+      - CONFIG_USERSPACE=n
+  drivers.uart.async_api.nocache_mem_dt.nucleo_f746zg:
+    harness: ztest
+    harness_config:
+      fixture: gpio_loopback
+    depends_on: gpio
+    platform_allow:
+      - nucleo_f746zg
+    extra_args:
+      - DTC_OVERLAY_FILE="boards/nucleo_f746zg.overlay;boards/nucleo_f746zg_nocachemem.overlay"
```
Comment (erwango) [inline tests/drivers/uart/uart_async_api/testcase.yaml:99]: Do you confirm that `boards/nucleo_f746zg.overlay` is not used if not specified ?

File: tests/drivers/uart/uart_async_api/testcase.yaml [hunk 5]
```diff
@@ -75,3 +75,36 @@ tests:
...
+  drivers.uart.async_api.nocache_mem_dt.nucleo_h723zg:
+    harness: ztest
+    harness_config:
+      fixture: gpio_loopback
+    depends_on: gpio
+    platform_allow:
+      - nucleo_h723zg
+    extra_args:
+      - DTC_OVERLAY_FILE="boards/nucleo_h723zg.overlay;boards/nucleo_h723zg_nocachemem.overlay"
+      - EXTRA_CONF_FILE=stm32_nocache-mem-dt.conf
```
Comment (erwango) [inline tests/drivers/uart/uart_async_api/testcase.yaml:110]: ditto

**Review by erwango — CHANGES_REQUESTED** [review]

File: tests/drivers/uart/uart_async_api/testcase.yaml (hunk 4 above)
Comment (ajarmouni-st) [reply tests/drivers/uart/uart_async_api/testcase.yaml:99]: yes

File: tests/drivers/uart/uart_async_api/src/test_uart_async.c [hunk 6]
```diff
@@ -1,11 +1,22 @@
...
 
+#if defined(CONFIG_DCACHE) && defined(CONFIG_NOCACHE_MEMORY)
+#define __NOCACHE	__attribute__ ((__section__(".nocache")))
+#define NOCACHE_MEM 1
+#elif defined(CONFIG_DCACHE) && defined(CONFIG_DT_DEFINED_NOCACHE)
+#define __NOCACHE	__attribute__ ((__section__(CONFIG_DT_DEFINED_NOCACHE_NAME)))
+#define NOCACHE_MEM 1
+#else
+#define NOCACHE_MEM 0
+#endif /* CONFIG_NOCACHE_MEMORY */
```
[the remark this answers is not in the document]
Comment (ajarmouni-st) [reply tests/drivers/uart/uart_async_api/src/test_uart_async.c:18]: interesting, will investigate.

File: tests/drivers/uart/uart_async_api/Kconfig (hunk 1 above)
Comment (ajarmouni-st) [reply tests/drivers/uart/uart_async_api/Kconfig:11]: > Perhaps it's time to generalize the DT NOCACHE? Considering its usage here, in the SPI tests, and potentially by applications.

@GeorgeCGV Will do it in another PR for all drivers as soon as this one is merged.

**Review by erwango — APPROVED** [review]

**Comment by erwango** [comment]
@dcpleung I'd like your opinion, specially on the test part, as this part is generic.

**Review by FRASTM — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]

**Pull request merged**